app/services/katello/applicability/applicable_content_helper.rb

```diff
@@ -38,19 +38,41 @@ def fetch_content_ids
-                                                  katello_installed_packages.id = \
-                                                  katello_host_installed_packages.installed_package_id WHERE \
-                                                  katello_repository_rpms.repository_id in \
-                                                  (#{self.bound_library_instance_repos.join(',')}) \
-                                                  and katello_host_installed_packages.host_id = \
-                                                  #{self.content_facet.host.id}").pluck(:id).uniq
+          # -> Include all non-modular rpms or rpms that exist within installed module streams
+          enabled_module_stream_ids = ::Katello::ModuleStream.
+            joins("inner join katello_available_module_streams on
+              katello_module_streams.name = katello_available_module_streams.name and
+                katello_module_streams.stream = katello_available_module_streams.stream").
+            joins("inner join katello_host_available_module_streams on
+              katello_available_module_streams.id = katello_host_available_module_streams.available_module_stream_id where
+                katello_host_available_module_streams.host_id = #{self.content_facet.host.id} and
+                katello_host_available_module_streams.status = 'enabled'").pluck(:id)
+
+          unless enabled_module_stream_ids.empty?
+            rpm_ids_in_enabled_module_streams = ::Katello::Rpm.
+              joins("inner join katello_module_stream_rpms on
+                katello_rpms.id = katello_module_stream_rpms.rpm_id where
+                  katello_module_stream_rpms.module_stream_id in (#{enabled_module_stream_ids.join(",")})").pluck(:id)
+          end
+
+          if rpm_ids_in_enabled_module_streams.empty?
+            rpms_in_modules_addendum = ""
+          else
+            rpms_in_modules_addendum = " or katello_rpms.id in (#{rpm_ids_in_enabled_module_streams.join(",")})"
+          end
+
+          return ::Katello::Rpm.
+            joins("INNER JOIN katello_repository_rpms ON
+              katello_rpms.id = katello_repository_rpms.rpm_id").
+            joins("INNER JOIN katello_installed_packages ON
+              katello_rpms.name = katello_installed_packages.name AND
+              katello_rpms.arch = katello_installed_packages.arch AND
+              katello_rpms.evr > katello_installed_packages.evr").
+            joins("INNER JOIN katello_host_installed_packages ON
+              katello_installed_packages.id = katello_host_installed_packages.installed_package_id WHERE
+                katello_repository_rpms.repository_id in (#{self.bound_library_instance_repos.join(',')}) and
+                katello_host_installed_packages.host_id = #{self.content_facet.host.id} and
+                (katello_rpms.modular = FALSE" + rpms_in_modules_addendum + ')').pluck(:id).uniq
```

Comment:
This is a tad simpler, and works in a single query (ignoring the bound repos being passed in):

```
          enabled_module_stream_ids = ::Katello::ModuleStream.
            joins("inner join katello_available_module_streams on
              katello_module_streams.name = katello_available_module_streams.name and
                  katello_module_streams.stream = katello_available_module_streams.stream").
            joins("inner join katello_host_available_module_streams on
              katello_available_module_streams.id = katello_host_available_module_streams.available_module_stream_id where
                katello_host_available_module_streams.host_id = #{self.content_facet.host.id} and
                katello_host_available_module_streams.status = 'enabled'").select(:id)

          query =  ::Katello::Rpm.
            joins("INNER JOIN katello_repository_rpms ON
              katello_rpms.id = katello_repository_rpms.rpm_id").
            joins("INNER JOIN katello_installed_packages ON
              katello_rpms.name = katello_installed_packages.name AND
              katello_rpms.arch = katello_installed_packages.arch AND
              katello_rpms.evr > katello_installed_packages.evr").
            joins("LEFT JOIN katello_module_stream_rpms ON
                   katello_rpms.id = katello_module_stream_rpms.rpm_id").
            joins("INNER JOIN katello_host_installed_packages ON
              katello_installed_packages.id = katello_host_installed_packages.installed_package_id").
            where("katello_repository_rpms.repository_id in (#{self.bound_library_instance_repos.join(',')})").
            where("katello_host_installed_packages.host_id = #{self.content_facet.host.id}").
            where("katello_module_stream_rpms.module_stream_id is null or katello_module_stream_rpms.module_stream_id in (:enabled_module_streams)", :enabled_module_streams => enabled_module_stream_ids).
            pluck(:id).uniq
```

Reply:
also, its probably time to split these 3 types of content out into their own method

Reply:
can you convert the other where() clauses to also use the prepared statement format like 3rd where clause there? Its safer, and you dont have to manually join()

Reply:
What do you mean by not having to manually join?

Reply:
i think my brain malfunctioned, i have no idea what i meant by that, but it looks like you got what i meant!

Reply:
ohhhh  the  array  join(',')  is what i meant!

Reply:
Ah that makes sense! Well that's taken care of then.